_posts/2024-06-14-remote-code-execution-engine.md

```diff
@@ -0,0 +1,1426 @@
+---
+title: "Distributed Remote Code Execution Engine"
+date: 2024-06-14
+header:
+    image: "https://res.cloudinary.com/dkoypjlgr/image/upload/f_auto,q_auto:good,c_auto,w_1200,h_300,g_auto,fl_progressive/v1715952116/blog_cover_large_phe6ch.jpg"
+tags: [scala,pekko,pekko-http,pekko-stream,pekko-cluster,docker,docker-compose,scala3]
+excerpt: "Practical guide to building the distributed remote code execution engine in Scala and Pekko"
+---
+
+_by [Anzori (Nika) Ghurtchumelia](https://github.com/ghurtchu)_
+
+## 1. Introduction
+
+After a long hiatus, I am back with renewed passion and energy, eager to delve deeper into the Scala ecosystem. This time, I am committed to building something tangible and useful with the tools available. Let's embark on this exciting journey of exploration and learning together.
+
+The greatest benefit of small side projects is the unique knowledge boost which can potentially be handy later in career.
+
+In this article we will attempt to build the remote code execution engine - the backend platform for websites such as [Hackerrank](https://hackerrank.com), [Leetcode](https://leetcode.com) and others.
+
+If, for some reason you're unfamiliar with the websites mentioned above, the basic usage flow is described below:
+- Client sends code
+- Backend runs it and responds with output
+
+There you go, sounds simple, right?
+
+Right, right... 
+
+Can you imagine how many things can go wrong here? It's the devil smirking in the corner, knowing, that the possibilities for failure are endless, however, we should address at least some of them.
+
+To give you a quick idea: a separate blog post can be written only about the security, not to mention scalability, extensibility and a few other compulsory properties to make it production ready.
+
+The goal isn't to build the best one, nor it is to compete with the existing ones. 
+
+Put simply, the goal of this project is to get familiar with `Pekko` and its modules such as `pekko-http`, `pekko-stream`, `pekko-cluster` and a few interesting concepts revolving around actor model concurrency, such as: 
+- cluster nodes and formation
+- cluster aware routers
+- remote worker actors
+- actor lifecycle and hierarchy
+- actor supervision strategies
+- actor location transparency
+- message serialization
+- and so on...
+
+Let's get started then, shall we?
+
+## 2. Project Structure
+
```

Comment:
Let's add the GitHub repo here and invite the reader to follow it, because they will likely not be able to replicate it from scratch.